modules/github-bots/sdk/github.go

```diff
@@ -83,6 +85,60 @@ func (c GitHubClient) Close(ctx context.Context) error {
 	return nil
 }
 
+// checkRateLimiting checks for github API rate limiting. It attempts to use
+// the returned Retry-After header to delay requests.
+//
+// Modified from https://github.com/wolfi-dev/wolfictl/blob/main/pkg/gh/github.go
+func checkRateLimiting(ctx context.Context, githubErr error) (bool, time.Duration) {
+	log := clog.FromContext(ctx)
+
+	// Default delay is 30 seconds
+	delay := time.Duration(30 * int(time.Second))
+	isRateLimited := false
+
+	// If GitHub returned an error of type RateLimitError, we can attempt to compute the next time to try the request again
+	// by reading its rate limit information
+	var rateLimitError *github.RateLimitError
+	if errors.As(githubErr, &rateLimitError) {
+		isRateLimited = true
+		retryAfter := time.Until(rateLimitError.Rate.Reset.Time)
+		delay = retryAfter
+		log.Infof("parsed retryAfter %d from GitHub rate limit error's reset time", retryAfter)
+	}
+
+	// If GitHub returned a Retry-After header, use its value, otherwise use the default
+	var abuseRateLimitError *github.AbuseRateLimitError
+	if errors.As(githubErr, &abuseRateLimitError) {
+		isRateLimited = true
+		if abuseRateLimitError.RetryAfter != nil {
+			if abuseRateLimitError.RetryAfter.Seconds() > 0 {
+				delay = *abuseRateLimitError.RetryAfter
+			}
+		}
+	}
+	return isRateLimited, delay
+}
+
+// handleGithubResponse handles the github response and error returned by most
+// API calls. It will handle checking for rate limiting as well as any errors
+// returned by the API.
+func handleGithubResponse(ctx context.Context, resp *github.Response, err error) error {
+	// if err is rate limit, delay and try again
+	githubErr := github.CheckResponse(resp.Response)
+	if githubErr != nil {
+		rateLimited, delay := checkRateLimiting(ctx, githubErr)
+		if !rateLimited {
+			return err
+		}
+		fmt.Printf("retrying after %v second delay due to rate limiting\n", delay.Seconds())
+		time.Sleep(delay)
```

Comment:
This makes me a little nervous tbh.

GH's rate limit quota resets hourly, so if we blow through it all at once then we could have jobs sleeping for 50m for the reset, costing money while they wait to time out and retry. Multiply that by potentially dozens of replicas, and it's not great.

I think we need a different approach to rate limits in the bot world, and we have some ideas.

Reply:
Good note @imjasonh

I don't need the rate limiting to complete my cycle item. Want me to just remove it from this PR and we can chat about the best way to do it moving forward?

Reply:
+1, lets separate the contentious rate-limiting parts and land this

Reply:
separated ✅